Error thrown: "java.lang.ArrayStoreException: java.lang.Byte"

Currently ignore test "flatMap" (in CollectionOperationsSpecification) throwing an error.

The script is 
OUTPUTS.flatMap({ (out: Box) => out.propositionBytes })(0) == 0.toByte

the error is:
Caused by: scalan.Base$StagingException: Cannot invoke method public abstract special.collection.Coll special.collection.Coll.flatMap(scala.Function1,scalan.RType) on object Coll(...)
...
Caused by: java.lang.ArrayStoreException: java.lang.Byte
	at scala.collection.mutable.ArrayBuilder$ofRef.$plus$eq(ArrayBuilder.scala:91)
	at scala.collection.mutable.ArrayBuilder$ofRef.$plus$eq(ArrayBuilder.scala:60)
	at scala.collection.generic.Growable.$anonfun$$plus$plus$eq$1(Growable.scala:62)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)

--- a/src/main/scala/sigmastate/eval/CostingRules.scala
+++ b/src/main/scala/sigmastate/eval/CostingRules.scala
@@ -544,18 +544,21 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
     }
 
     def flatMap[B](fC: RCosted[T => Coll[B]]): RCostedColl[B] = {
-      val f = fC.value
-      f match {
+      val fV = fC.value
+      fV match {
         // Pattern: xs.flatMap(x => x.property)
         case Def(Lambda(l,_,_,Def(mc @ MethodCall(x, m, Nil, _)))) if x == l.x =>
-          val sObj = asSizeColl(obj.size)
-          val sizeF = asCostedFunc[T, Coll[B]](fC).sliceSize
-          val sizes: RColl[Size[B]] = sObj.sizes.flatMap(fun { s: RSize[T] =>
+          val cfC = asCostedFunc[T, Coll[B]](fC)
+          val calcF = cfC.sliceCalc
+          val sizeF = cfC.sliceSize
+          val costF = cfC.sliceCost
+          val xs = asCostedColl(obj)
+          val vals = xs.values.flatMap(calcF)
+          val sizes: RColl[Size[B]] = xs.sizes.flatMap(fun { s: RSize[T] =>
             asSizeColl(sizeF(s)).sizes
           })
-          val values = obj.value.flatMap(f)
-          val costs = colBuilder.replicate(sizes.length, 0)
-          RCCostedColl(values, costs, sizes, opCost(values, costOfArgs, costOf(method)))
+          val costs = xs.costs.zip(xs.sizes).map(costF)
+          RCCostedColl(vals, costs, sizes, opCost(vals, costOfArgs, costOf(method)))
         case _ =>
           !!!(s"Unsupported lambda in flatMap: allowed usage `xs.flatMap(x => x.property)`")
       }

--- a/src/main/scala/sigmastate/eval/CostingRules.scala
+++ b/src/main/scala/sigmastate/eval/CostingRules.scala
@@ -543,20 +543,14 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
       RCCostedColl(vals, costs, sizes, opCost(vals, costOfArgs, costOf(method)))
     }
 
-    def getSizePropertyMethod[B](mc: MethodCall): RSize[T] => RColl[Size[B]] = {
-      ???
-    }
-
     def flatMap[B](fC: RCosted[T => Coll[B]]): RCostedColl[B] = {
-      val sizeF = asCostedFunc[T, Coll[B]](fC).sliceSize
       val f = fC.value
       f match {
         // Pattern: xs.flatMap(x => x.property)
         case Def(Lambda(l,_,_,Def(mc @ MethodCall(x, m, Nil, _)))) if x == l.x =>
           val sObj = asSizeColl(obj.size)
+          val sizeF = asCostedFunc[T, Coll[B]](fC).sliceSize
           val sizes: RColl[Size[B]] = sObj.sizes.flatMap(fun { s: RSize[T] =>
-//            val sizeProp = getSizePropertyMethod[B](mc)
-//            sizeProp(s)
             asSizeColl(sizeF(s)).sizes
           })
           val values = obj.value.flatMap(f)

--- a/src/main/scala/sigmastate/eval/CostingRules.scala
+++ b/src/main/scala/sigmastate/eval/CostingRules.scala
@@ -548,18 +548,20 @@ trait CostingRules extends SigmaLibrary { IR: RuntimeCosting =>
     }
 
     def flatMap[B](fC: RCosted[T => Coll[B]]): RCostedColl[B] = {
+      val sizeF = asCostedFunc[T, Coll[B]](fC).sliceSize
       val f = fC.value
       f match {
         // Pattern: xs.flatMap(x => x.property)
         case Def(Lambda(l,_,_,Def(mc @ MethodCall(x, m, Nil, _)))) if x == l.x =>
           val sObj = asSizeColl(obj.size)
           val sizes: RColl[Size[B]] = sObj.sizes.flatMap(fun { s: RSize[T] =>
-            val sizeProp = getSizePropertyMethod[B](mc)
-            sizeProp(s)
+//            val sizeProp = getSizePropertyMethod[B](mc)
+//            sizeProp(s)
+            asSizeColl(sizeF(s)).sizes
           })
           val values = obj.value.flatMap(f)
           val costs = colBuilder.replicate(sizes.length, 0)
-          RCCostedColl(values, costs, sizes, opCost(obj.value, costOfArgs, costOf(method)))
+          RCCostedColl(values, costs, sizes, opCost(values, costOfArgs, costOf(method)))
         case _ =>
           !!!(s"Unsupported lambda in flatMap: allowed usage `xs.flatMap(x => x.property)`")
       }

--- a/src/main/scala/sigmastate/eval/Evaluation.scala
+++ b/src/main/scala/sigmastate/eval/Evaluation.scala
@@ -375,7 +375,7 @@ trait Evaluation extends RuntimeCosting { IR =>
             val dataRes = obj.elem match {
               case _: CollElem[_, _] => mc match {
                 case CollMethods.flatMap(xs, f) =>
-                  val newMC = mc.copy(args = mc.args :+ f.elem.eRange)(mc.selfType, mc.isAdapterCall)
+                  val newMC = mc.copy(args = mc.args :+ f.elem.eRange.eItem)(mc.selfType, mc.isAdapterCall)
                   invokeUnlifted(obj.elem, newMC, dataEnv)
                 case _ =>
                   invokeUnlifted(obj.elem, mc, dataEnv)

--- a/src/main/scala/sigmastate/utxo/CostTable.scala
+++ b/src/main/scala/sigmastate/utxo/CostTable.scala
@@ -134,6 +134,7 @@ object CostTable {
     ("SizeOf", "(Coll[IV]) => Int", collLength),
     ("ByIndex", "(Coll[IV],Int) => IV", collByIndex),
     ("SCollection$.map", "(Coll[IV],(IV) => OV) => Coll[OV]", collToColl),
+    ("SCollection$.flatMap", "(Coll[IV],(IV) => Coll[OV]) => Coll[OV]", collToColl),
     ("SCollection$.indexOf_per_kb", "(Coll[IV],IV,Int) => Int", collToColl),
     ("SCollection$.segmentLength", "(Coll[IV],(IV) => Boolean,Int) => Int", collToColl),
     ("SCollection$.indexWhere", "(Coll[IV],(IV) => Boolean,Int) => Int", collToColl),

--- a/src/test/scala/sigmastate/CostingSpecification.scala
+++ b/src/test/scala/sigmastate/CostingSpecification.scala
@@ -209,6 +209,9 @@ class CostingSpecification extends SigmaTestingData {
     cost(s"{ $coll.filter({ (b: Box) => b.value > 1L }).size > 0 }") shouldBe
       (lambdaCost + accessBox + extractCost + GTConstCost + selectField +
         (accessBox + comparisonCost) * tx.outputs.length + collToColl + LengthGTConstCost)
+    cost(s"{ $coll.flatMap({ (b: Box) => b.propositionBytes }).size > 0 }") shouldBe
+      (lambdaCost + accessBox + extractCost + selectField +
+        accessBox  * tx.outputs.length + collToColl + LengthGTConstCost)
   }
 
   property("Option operations cost") {

--- a/src/test/scala/sigmastate/utxo/CollectionOperationsSpecification.scala
+++ b/src/test/scala/sigmastate/utxo/CollectionOperationsSpecification.scala
@@ -440,9 +440,7 @@ class CollectionOperationsSpecification extends SigmaTestingCommons {
     assertProof(code, expectedPropTree, outputBoxValues)
   }
 
-  //TODO: related to https://github.com/ScorexFoundation/sigmastate-interpreter/issues/423
-  // TODO costing rule in CollCoster
-  ignore("flatMap") {
+  property("flatMap") {
     assertProof("OUTPUTS.flatMap({ (out: Box) => out.propositionBytes })(0) == 0.toByte",
       EQ(
         ByIndex(

--- a/src/test/scala/special/sigma/SigmaDslTest.scala
+++ b/src/test/scala/special/sigma/SigmaDslTest.scala
@@ -438,6 +438,7 @@ class SigmaDslTest extends PropSpec
     val coll = ctx.OUTPUTS
     val eq = EqualityChecker(coll)
     eq({ (x: Coll[Box]) => x.filter({ (b: Box) => b.value > 1 }) })("{ (x: Coll[Box]) => x.filter({(b: Box) => b.value > 1 }) }")
+    eq({ (x: Coll[Box]) => x.flatMap({ (b: Box) => b.propositionBytes }) })("{ (x: Coll[Box]) => x.flatMap({(b: Box) => b.propositionBytes }) }")
   }
 
   property("Option methods equivalence") {

